calico/etcddriver/driver.py

```diff
@@ -305,6 +308,16 @@ def _resync_and_merge(self):
                 self._resync_requested = False
         _log.info("Stop event set, exiting resync loop.")
 
+    def _rotate_etcd_url(self):
```

Comment:
I didn't really like passing the URLs (others needed too) through to the watcher - there's a lack of thread local storage, so it's a bunch of parameters, and having the different threads run with different values is a bit unpleasant...

I was tempted to modify the watcher to always die instead of rotating, and have the main thread do the rotate in response, but I've gone with the lock.  The lock is used in more places than I'd ideally like, but it's not recursive and there are no other locks, so deadlocks shouldn't be a problem.